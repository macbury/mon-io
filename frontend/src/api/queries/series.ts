import gql from 'graphql-tag'

import ViewSeries from '../fragments/ViewSeries'
import ViewCategory from '../fragments/ViewCategory'
import ViewLocation from '../fragments/ViewLocation'
import ViewCurrency from '../fragments/ViewCurrency'

export default gql`
  ${ViewSeries}
  ${ViewCategory}
  ${ViewCurrency}
  ${ViewLocation}

  query series {
    series {
      nodes {
        ...ViewSeries
        blueprint {
          notes
          kind
          category {
            ...ViewCategory
          }

          location {
            ...ViewLocation
          }

          amount {
            cents
            currency {
              ...ViewCurrency
            }
          }

          exchangedAmount {
            currency {
              ...ViewCurrency
            }
            cents
          }
        }
      }
    }
  }
`