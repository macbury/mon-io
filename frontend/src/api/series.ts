import { ApiBase } from './types'
import moment, { Moment } from 'moment-timezone'

import {
  MutationIgnoreRecurrenceArgs,
  MutationCreateRecurrenceArgs,
  MutationUpdateRecurrenceArgs,
  CreateRecurrenceInput,
  UpdateRecurrenceInput,
  Mutation,
  Query,
  QueryGetSeriesArgs,
  QueryPlannedTransactionsArgs
} from './graphql'

import createRecurrenceMutation from './mutations/createRecurrence'
import updateRecurrenceMutation from './mutations/updateRecurrence'
import ignoreRecurrenceMutation from './mutations/ignoreRecurrence'
import plannedTransactionsQuery from './queries/getPlannedTransactions'
import getSeriesQuery from './queries/getSeries'
import seriesQuery from './queries/series'
import getCalendarUrlQuery from './queries/getCalendarUrl'

export default class SeriesApi extends ApiBase {
  public async create(input : CreateRecurrenceInput) {
    try {
      await this.client.mutate<Mutation, MutationCreateRecurrenceArgs>({
        mutation: createRecurrenceMutation,
        variables: { input }
      })

      return { errors: [] }
    } catch (error) {
      return { errors: [error.toString()] }
    }
  }

  public async update(input : UpdateRecurrenceInput) {
    try {
      const { data: { updateRecurrence } } = await this.client.mutate<Mutation, MutationUpdateRecurrenceArgs>({
        mutation: updateRecurrenceMutation,
        variables: { input }
      })

      return updateRecurrence
    } catch (error) {
      return { errors: [error.toString()] }
    }
  }

  public async getCalendarUrl() {
    try {
      const { data } = await this.client.query<Query>({
        query: getCalendarUrlQuery
      })

      return (data as any)?.me?.settings?.calendarUrl
    } catch (error) {
      return null
    }
  }

  public async find(id : string) {
    const { data: { getSeries } } =  await this.client.query<Query, QueryGetSeriesArgs>({
      query: getSeriesQuery,
      variables: { id }
    })

    return getSeries
  }

  public async ignoreDate(seriesId : string, date : string | Moment) {
    const atDate = moment(date).format('YYYY-MM-DD')
    try {
      const { data: { ignoreRecurrence } } = await this.client.mutate<Mutation, MutationIgnoreRecurrenceArgs>({
        mutation: ignoreRecurrenceMutation,
        variables: { input: { seriesId, atDate } }
      })

      return { errors: ignoreRecurrence.errors }
    } catch (error) {
      return { errors: [error.toString()] }
    }
  }

  public async all() {
    const { data: { series } } =  await this.client.query<Query>({
      query: seriesQuery
    })

    return series?.nodes
  }

  public async getPlannedTransactions(date : string | Moment, all = false) {
    const month = moment(date).format('YYYY-MM-DD')

    const { data: { plannedTransactions } } =  await this.client.query<Query, QueryPlannedTransactionsArgs>({
      query: plannedTransactionsQuery,
      variables: { month, all }
    })

    return plannedTransactions
  }
}
