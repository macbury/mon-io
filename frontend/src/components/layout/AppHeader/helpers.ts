import { AppHeaderProp } from './types'

export function getTitleFromNavigation(props : AppHeaderProp) : string {
  let title = props.title
  if (!title) {
    title = props.scene?.descriptor?.options?.title
  }
  return title
}