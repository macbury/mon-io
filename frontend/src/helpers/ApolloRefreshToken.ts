import SessionStore from '../stores/SessionStore'

export default class ApolloRefreshToken {
  private refreshing : boolean;
  private resolvers : Array<Function> = []
  private sessions : SessionStore

  constructor(sessions : SessionStore) {
    this.sessions = sessions
  }

  public refreshToken() : Promise<void> {
    return new Promise<void>((resolve) => {
      this.resolvers.push(resolve)

      if (!this.refreshing) {
        this.sessions.refreshServerToken().then(this.onRefreshEnd).catch(this.onRefreshEnd)
        this.refreshing = true
      }
    })
  }

  private onRefreshEnd = () => {
    this.refreshing = false
    this.resolvers.forEach((resolve) => resolve())
    this.resolvers = []
  }
}