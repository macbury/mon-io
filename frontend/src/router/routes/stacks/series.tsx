import React from 'react'
import { DefaultTheme } from 'styled-components/native'
import createStack from './createStack'

import SeriesScreen from '../../../screens/SeriesScreen'
import EditTransactionScreen from '../../../screens/Transactions/EditTransactionScreen'
import EditSeriesScreen from '../../../screens/EditSeriesScreen'
import PreviewReceiptScreen from '../../../screens/PreviewReceiptScreen'
import NewTransactionScreen from '../../../screens/Transactions/NewTransactionScreen'

export default function createSeriesStack(theme : DefaultTheme) {
  return createStack({
    series: {
      screen: SeriesScreen,
      path: ':date',
      params: {
        date: 'today'
      },
      navigationOptions: {
        forceShowHeader: true
      }
    },
    editSeries: {
      screen: EditSeriesScreen,
      path: ':date/:seriesId/edit/:updateType'
    },
    newPlannedTransaction: {
      screen: NewTransactionScreen,
      path: 'plan/:seriesId'
    },
    newSeries: {
      screen: NewTransactionScreen,
      path: ':recurrence/:createdAt/new'
    },
    editTransaction: {
      screen: EditTransactionScreen,
      path: ':transactionId/transaction'
    },
    preview: {
      screen: PreviewReceiptScreen,
      path: ':receiptId/preview'
    },
  }, theme)
}