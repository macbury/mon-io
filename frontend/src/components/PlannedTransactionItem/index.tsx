import React, { useMemo, useCallback } from 'react'
import moment from 'moment-timezone'
import { useTranslation } from 'react-i18next'
import styled, { DefaultTheme } from 'styled-components/native'
import { Text } from 'react-native-paper'
import { NavigationNavigateAction } from 'react-navigation'
import { PlannedTransaction, Series } from '../../api/graphql'
import Link from '../Link'
import CategoryIcon from '../CategoryIcon'
import Currency from '../Currency'

export enum DetailsType {
  NextDate,
  Recurrence
}

interface IStyleProp {
  bigger?: boolean
  theme?: DefaultTheme
}
export interface IPlannedTransactionItemProps {
  plannedTransaction: PlannedTransaction
  series: Series
  bigger?: boolean
  hidePlanned?: boolean
  action: NavigationNavigateAction
  type?: DetailsType
  onActionsShow(series : Series, date : string)
}

const Inner = styled.View<IStyleProp>`
  display: flex;
  flex-direction: row;
  padding: ${({ bigger }) => bigger ? '13px 15px' : '10px 15px'};
  align-content: center;
  border-bottom-width: 1px;
  border-bottom-color: ${(prop) => prop.theme.itemBorderColor};
  align-items: center;
`

const Icon = styled(CategoryIcon)`
  margin-bottom: 0px;
  margin-right: 15px;
`

const Details = styled.View`
  display: flex;
  flex-direction: column;
  flex: 1;
`

const DetailsRow = styled.View`
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  margin-top: 0px;
`

interface ICategoryNameProps {
  color: string
}

const CategoryName = styled(Text)`
  font-weight: bold;
  font-size: 16px;
  color: ${({ color } : ICategoryNameProps) => color};
`

const Date = styled(Text)`
  opacity: 0.4;
  margin-left: 5px;
`

const Notes = styled(Text)`
  font-style: italic;
  margin-top: 2px;
`

export default function PlannedTransactionItem(props : IPlannedTransactionItemProps) {
  const {
    bigger,
    action,
    hidePlanned,
    plannedTransaction: {
      category,
      amount,
      date,
      notes
    },
    series,
    type,
    onActionsShow
  } = props

  const { t } = useTranslation()
  const showPlannedActionSheet = useCallback(() => onActionsShow(series, date), [onActionsShow, series, date])

  const formattedDate = useMemo(() => {
    if (hidePlanned) {
      return ''
    }

    return t(
      'planned_transaction.at_date',
      { date: moment(date).format('Do MMM YYYY') }
    )
  }, [date, t, hidePlanned])

  return (
    <Link ripple action={action} onLongPress={showPlannedActionSheet}>
      <Inner bigger={bigger}>
        <Icon
          name={category.icon}
          color={category.color}
          size={48} />
        <Details>
          <DetailsRow>
            <CategoryName color={category.color}>{category.name}</CategoryName>
            {!hidePlanned && type !== DetailsType.Recurrence && <Date>{formattedDate}</Date>}
          </DetailsRow>

          {notes.length > 0 && <Notes numberOfLines={1}>{notes}</Notes>}
        </Details>

        <Currency
          amount={amount}
          size={18} />
      </Inner>
    </Link>
  )
}