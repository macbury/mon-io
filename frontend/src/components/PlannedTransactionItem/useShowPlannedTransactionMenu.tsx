import React from 'react'
import { useCallback } from 'react'
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons'
import { useTheme } from 'styled-components/native'
import { useTranslation } from 'react-i18next'
import { Series } from '../../api/graphql'
import useActionSheet from '../../helpers/useActionSheet'

interface IUseShowPlannedTransactionMenuArgs {
  onIgnorePlannedTransaction(seriesId : string, date : string)
  onEditPlannedTransaction(seriesId : string, date : string)
}

export default function useShowPlannedTransactionMenu(opts : IUseShowPlannedTransactionMenuArgs) {
  const showActionSheetWithOptions = useActionSheet()
  const { t } = useTranslation()
  const theme = useTheme()

  return useCallback((series: Series, date? : string) => {
    const {
      onIgnorePlannedTransaction,
      onEditPlannedTransaction
    } = opts

    showActionSheetWithOptions({
      icons: [
        <MaterialCommunityIcons color={theme.colors.text} size={24} name='calendar-remove' />,
        <MaterialCommunityIcons color={theme.colors.text} size={24} name="calendar-edit" />,
        <MaterialCommunityIcons color={theme.colors.text} size={24} name="cancel" />
      ],
      options: [
        t('actions.planned_transaction.ignore'),
        t('actions.planned_transaction.edit_series'),
        t('actions.planned_transaction.cancel')
      ],
      cancelButtonIndex: 2,
    }, (option) => {
      switch(option) {
        case 0: // Ignore
          onIgnorePlannedTransaction(series.id, date)
        break;

        case 1:
          onEditPlannedTransaction(series.id, date)
        break
      }
    })
  }, opts)
}