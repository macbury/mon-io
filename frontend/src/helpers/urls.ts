import { Recurrence } from './../api/graphql';
import moment, { Moment } from 'moment-timezone'
import { NavigationActions, NavigationNavigateAction } from 'react-navigation'
import { TransactionCategoryKind, RecurrenceUpdateMode } from '../api/graphql'

export function homePath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'home' })
}

export function receiptsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'receipts' })
}

export function mapPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'map' })
}

export function authPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'auth' })
}

export function importsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'imports' })
}

export function editImportPath(importId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'newImport', params: { importId } })
}

export function settingsApiExplorerPath(refreshTokenId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'apiExplorer', params: { refreshTokenId } })
}

export function categoriesPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'categories' })
}

export function newCategoryPath(kind: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'newCategory', params: { kind } })
}

export function editCategoriesPath(categoryId: string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'editCategory', params: { categoryId } })
}

export function scanPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'scan' })
}

export function changeNewTransactionCategoryPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'changeNewTransactionCategory' })
}

export function signInPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'auth' })
}

export function seriesPath(date?: Moment) : NavigationNavigateAction {
  const params = date ? { date: date.format('YYYY-MM-DD') } : {}
  return NavigationActions.navigate({ routeName: 'series', params })
}

export function chartsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'charts' })
}

export function editSeriesPath(seriesId : string, updateType: RecurrenceUpdateMode, occuredAt?: Moment | string) : NavigationNavigateAction {
  let date = occuredAt ? moment(occuredAt).format('YYYY-MM-DD') : null
  return NavigationActions.navigate({ routeName: 'editSeries', params: { seriesId, date, updateType } })
}

export function editCategoryBudgetPath(budgetId : string, categoryId : string, kind: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'editCategoryBudget', params: { budgetId, categoryId, kind } })
}

export function budgetPath(date : Moment) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'budget', params: { date: date.format('YYYY-MM') } })
}

export function summaryPath(date : Moment, type?: TransactionCategoryKind) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'summary', params: { date: date.format('YYYY-MM'), type } })
}

export function transactionsPath(date? : Moment, type?: TransactionCategoryKind, importId?: string) : NavigationNavigateAction {
  return NavigationActions.navigate({
    routeName: 'transactions',
    params: {
      date: date?.format('YYYY-MM'),
      type,
      importId
    }
  })
}

export function previewReceiptPath(receiptId : string) : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'preview', params: { receiptId } })
}

export function settingsPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settings' })
}

export function settingsRefreshTokensPath() : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsRefreshTokens' })
}

export function settingsAuthorizeDevicePath()  : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsQuickLogin' })
}

export function settingsDownloadMobileAppPath()  : NavigationNavigateAction {
  return NavigationActions.navigate({ routeName: 'settingsDownloadMobileApp' })
}

export function newSummaryTransactionPath(categoryId: string, createdAt: string, kind: TransactionCategoryKind) {
  return NavigationActions.navigate({
    routeName: 'newSummaryTransaction',
    params: {
      categoryId, createdAt, kind
    }
  })
}

export function newBillTransactionPath(receiptId: string) {
  return NavigationActions.navigate({ routeName: 'newBillTransaction', params: { receiptId } })
}

export function newPlannedTransactionPath(seriesId: string, createdAt: string) {
  return NavigationActions.navigate({ routeName: 'newPlannedTransaction', params: { createdAt, seriesId } })
}

export function newSeriesPath(createdAt: string) {
  return NavigationActions.navigate({ routeName: 'newSeries', params: { createdAt, recurrence: Recurrence.Once } })
}

export function editSummaryTransactionPath(transactionId : string) {
  return NavigationActions.navigate({ routeName: 'editSummaryTransaction', params: { transactionId } })
}

export function editTransactionPath(transactionId : string) {
  return NavigationActions.navigate({ routeName: 'editTransaction', params: { transactionId } })
}