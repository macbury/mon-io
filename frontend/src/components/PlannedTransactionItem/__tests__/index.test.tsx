import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import { newPlannedTransactionPath } from '../../../helpers/urls'
import { series as seriesBuilder } from '../../../test/factories/series'

import PlannedTransactionItem, { IPlannedTransactionItemProps } from '../index'

it('renders correctly', () => {
  const series = seriesBuilder.build()

  const props : IPlannedTransactionItemProps = {
    action: newPlannedTransactionPath('seriesId', 'isoDate'),
    series,
    onActionsShow: jest.fn(),
    plannedTransaction: series.blueprint as any
  }

  const component = shallow(<PlannedTransactionItem {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})