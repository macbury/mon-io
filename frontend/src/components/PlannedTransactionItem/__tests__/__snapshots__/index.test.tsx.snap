// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Link
  action={
    Object {
      "params": Object {
        "createdAt": "isoDate",
        "seriesId": "seriesId",
      },
      "routeName": "newPlannedTransaction",
      "type": "Navigation/NAVIGATE",
    }
  }
  onLongPress={[Function]}
  ripple={true}
>
  <Styled(View)>
    <Styled(CategoryIcon)
      color="#fff"
      name="MaterialIcons:repeat"
      size={48}
    />
    <Styled(View)>
      <Styled(View)>
        <Styled(withTheme(Text))
          color="#fff"
        >
          Category
        </Styled(withTheme(Text))>
        <Styled(withTheme(Text))>
          next at 12th Jan 2020
        </Styled(withTheme(Text))>
      </Styled(View)>
      <Styled(withTheme(Text))
        numberOfLines={1}
      >
        random transaction
      </Styled(withTheme(Text))>
    </Styled(View)>
    <Currency
      amount={
        Object {
          "cents": 1000,
          "currency": Object {
            "id": "PLN",
            "isoCode": "PLN",
            "name": "Polish zloty",
            "subunitToUnit": 100,
            "symbol": "zÅ‚",
          },
        }
      }
      size={18}
    />
  </Styled(View)>
</Link>
`;
